packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
     }
   });
   return plugins;
 }
 
+void _addPluginsToPlatform(List<Plugin> plugins, FlutterProjectPlatform platform, Map<String, dynamic> pluginsMap) {
+  if (!platform.existsSync()) {
+    return;
+  }
+  final List<Map<String,dynamic>> list = platform.pluginsList(plugins);
+  if (list.isNotEmpty) {
+    pluginsMap[platform.pluginConfigKey] = list;
+  }
+}
+
+/// Writes the .flutter-plugins.json file based on the list of plugins.
+/// If there aren't any plugins, then the files aren't written to disk. The resulting
+/// file looks something like this:
+/// {
+///     "_info": "// This is a generated file; do not edit or check into version control.",
+///     "plugins": {
+///         "ios": [
+///             {
+///                 "name": "plugin-a",
+///                 "path": "/path/to/plugin/pubspec",
+///                 "dependencies": []
+///             },
+///             {
+///                 "name": "plugin-b",
+///                 "path": "/path/to/plugin/pubspec",
+///                 "dependencies": [
+///                     "plugin-c",
+///                     "plugin-d"
+///                 ]
+///             }
+///         ],
+///         "macos": [
+///             {
+///                 "name": "plugin-a",
+///                 "path": "/path/to/plugin/pubspec",
+///                 "dependencies": [
+///                    plugin-b,
+///                 ]
+///             },
+///             {
+///                 "name": "plugin-b",
+///                 "path": "/path/to/plugin/pubspec",
+///                "dependencies": [
+///                     "plugin-c"
+///                 ]
+///             }
+///         ],
+///     }
+/// }
+///
+/// Finally, returns [true] if .flutter-plugins.json has changed,
+/// otherwise returns [false].
+bool _writeFlutterPluginsList(FlutterProject project, List<Plugin> plugins) {
+  const String info = 'This is a generated file; do not edit or check into version control.';
+
+  final Map<String, dynamic> result = <String, dynamic> {};
+  result['_info'] = '// $info';
+
+  final Map<String, dynamic> pluginsMap = <String, dynamic>{};
+  _addPluginsToPlatform(plugins, project.ios, pluginsMap);
+  _addPluginsToPlatform(plugins, project.android, pluginsMap);
+  _addPluginsToPlatform(plugins, project.macos, pluginsMap);
+  _addPluginsToPlatform(plugins, project.linux, pluginsMap);
+  _addPluginsToPlatform(plugins, project.windows, pluginsMap);
+  _addPluginsToPlatform(plugins, project.web, pluginsMap);
+
+  result['plugins'] = pluginsMap;
+
+  final File pluginsFile = project.flutterPluginsJsonFile;
+  final String oldPluginFileContent = _readFileContent(pluginsFile);
+  final String pluginFileContent = json.encode(result);
+  if (pluginsMap.isNotEmpty) {
+    pluginsFile.writeAsStringSync(pluginFileContent, flush: true);
+  } else {
+    if (pluginsFile.existsSync()) {
+      pluginsFile.deleteSync();
+    }
+  }
+
+  return oldPluginFileContent != _readFileContent(pluginsFile);
```

Comment:
Since there's encoding/decoding going on now, it would be nice to make this a little bit more efficient. You don't need to `json.encode` above if `pluginsMap.isEmpty`, and you don't need to do the file readback since you already have `pluginFileContent`. There's already branching for empty vs. non-empty, so being more efficient here would add minimal complexity. Something like:

```
  final File pluginsFile = project.flutterPluginsJsonFile;
  final String oldPluginFileContent = _readFileContent(pluginsFile);
  final String pluginFileContent = pluginsMap.isNotEmpty ? json.encode(result) : null;
  if (pluginFileContent != null) {
    pluginsFile.writeAsStringSync(pluginFileContent, flush: true);
  } else {
    if (pluginsFile.existsSync()) {
      pluginsFile.deleteSync();
    }
  }

  return pluginFileContent != oldPluginFileContent;
```

Reply:
I moved the check for empty plugins to the top. Feels a bit cleaner to me. WDYT?